### PR TITLE
Take more advantage of angle offset learner

### PR DESCRIPTION
**Pull request by Gernby**
After much experimentation and testing, I have finally come to the conclusion that LiveParameters are not only good, but they are also under-utilized.  

I've greatly improved the overall user experience across all makes / models in my fully customized branches with this enhancement, and the Tesla fork seems to benefit as well ([PR here](https://github.com/SippieCup/openpilot/pull/11)).

### CODE REVIEW

File: selfdrive/controls/lib/latcontrol_indi.py [hunk 1]
```diff
@@ -64,6 +64,8 @@ def update(self, active, v_ego, angle_steers, angle_steers_rate, steer_override,
     else:
       self.angle_steers_des = path_plan.angleSteers
       self.rate_steers_des = path_plan.rateSteers
+      if not steer_override:
```
Comment (Gernby) [inline selfdrive/controls/lib/latcontrol_indi.py]: The fast angle offset value becomes disruptive during steer_override, so it's best to NOT update the bias during the override

File: selfdrive/controls/lib/pathplanner.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def update(self, sm, CP, VM):
     active = sm['controlsState'].active
 
     angle_offset_average = sm['liveParameters'].angleOffsetAverage
-    angle_offset_bias = sm['controlsState'].angleModelBias + angle_offset_average
```
Comment (Gernby) [inline selfdrive/controls/lib/pathplanner.py]: The angleModelBias value is hard-coded to 0 in controlsd, so there's no value in this parameter

File: selfdrive/controls/lib/pathplanner.py [hunk 3]
```diff
@@ -73,7 +73,7 @@ def update(self, sm, CP, VM):
     self.p_poly = list(self.MP.p_poly)
 
     # account for actuation delay
-    self.cur_state = calc_states_after_delay(self.cur_state, v_ego, angle_steers - angle_offset_average, curvature_factor, VM.sR, CP.steerActuatorDelay)
```
Comment (Gernby) [inline selfdrive/controls/lib/pathplanner.py]: The slow angle offset isn't as "correct" as the fast angle offset here.

**Comment by pd0wm** [comment]
I'm aware that using the fast angle offset will improve driving in some scenarios. However, I found that enabling it on a straight highway introduces slight ping pong. I think this is because the fast angle offset learner is a bit too noisy.

Unless there are no regressions by switching to the fast angle offset learner, I'm not convinced it's ready for general use. Especially because for most people highway driving is the main use case of openpilot.

**Comment by Gernby** [comment]
Thanks for the response.  I did receive some feedback that it was overly active in some cases,  However, I found that a simple rate limit is a very effective noise filter.  Please see the latest commit.

**Comment by pd0wm** [comment]
The noise probably orignates from the yaw rate output of the kalman filter in `locationd_yawrate.cc`. If it's too noisy I should retune that to make it more stable. I'll be on holiday for a bit, but when I'm back I'll see if I can make the fast learner more stable.

**Comment by Gernby** [comment]
I haven't updated the PR branch yet, but I wanted to share the results from recent improvements to the noise filter.  I've enhanced the filter so that it is more restrictive when A) curvature is low and / or B) desired rate is low.  When curvature AND rate are both low, noise is heavily filtered.

image

**Review by pmod90 — APPROVED** [review]

**Comment by rbiasini** [comment]
we shipped fast angle offset learner in 0.6.4. Thanks!

**Pull request closed without merging**